internal/provider/kubernetes/gateway.go

```diff
@@ -280,6 +413,117 @@ func (r *gatewayReconciler) envoyServiceForGateway(ctx context.Context, gateway
 	return svc, nil
 }
 
+// otherGatewaysRefSecret returns an error if a managed Gateway other than the provided
+// gateway references the provided secret, or if an error is encountered while checking.
+func (r *gatewayReconciler) otherGatewaysRefSecret(ctx context.Context, gateway *gwapiv1b1.Gateway, secret *corev1.Secret) error {
+	if gateway == nil || secret == nil {
+		return nil
+	}
+	gateways := &gwapiv1b1.GatewayList{}
+	if err := r.client.List(ctx, gateways); err != nil {
+		return fmt.Errorf("error listing gatewayclasses: %v", err)
+	}
+	for i := range gateways.Items {
+		gw := gateways.Items[i]
+		if gw.Namespace != gateway.Namespace && gw.Namespace != gateway.Name && r.hasMatchingController(&gw) {
+			secrets, err := r.secretsForGateway(ctx, &gw)
+			if err != nil {
+				return err
+			}
+			for _, s := range secrets {
+				if s.Namespace == secret.Namespace && s.Name == secret.Name {
+					return fmt.Errorf("secret %s/%s referenced", s.Namespace, s.Name)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// secretsForGateway returns the Secrets referenced by the provided gateway listeners.
+func (r *gatewayReconciler) secretsForGateway(ctx context.Context, gateway *gwapiv1b1.Gateway) ([]corev1.Secret, error) {
```

Comment:
would help if there was a test for `secretsForGateway` added in this or follow up PRs